### ISSUE TITLE
Compatibility with Advanced Markers

### ISSUE THREAD

**Issue by MathieuAlphamosa** [googlemaps/js-markerclusterer#508]
Advanced Markers is still in beta but js-markerclusterer should plan compatibility with it.  
https://developers.google.com/maps/documentation/javascript/advanced-markers/overview

I didn't go far with it, first problem is that the Advanced Marker class don't have a getPosition function.

**Comment by amuramoto** [comment]
Thanks for opening this issue! I am in progress on this [here](https://github.com/googlemaps/js-markerclusterer/commit/4f23dd1304f72cc6b86989dbd168f093a5ff7240). The goal is to have Advanced Markers support available by GA launch.

**Comment by nebulousGirl** [comment]
I ran in the same issue. AdvancedMarkers are missing a bunch of methods that the clusterer relies on.

**Comment by bdcoder2** [comment]
For anyone that is interested, [this grid clusterer](https://github.com/bdcoder2/grid_clusterer) now support both &quot;classic&quot; and &quot;advanced&quot; markers.

**Comment by nemanjacosovic** [comment]
Hey @amuramoto, can we get a status update?

**Comment by amuramoto** [comment]
Right now I've got type checks passing and the cluster icons themselves implemented as Adv Markers when available. Haven't yet gotten to the internals of what happens when an array of Adv Markers is passed in (I'm sure it will break the world, just not sure how badly yet). 

There's a few other higher priority fires I'm trying to put out right now, so work on this is delayed but still planning to align this with the GA launch of Adv Markers

**Comment by nemanjacosovic** [comment]
Thank you @amuramoto can we do anything to assist? I'm volunteering as a tribute.

**Comment by Imtinan1996** [comment]
Hi Guys, just wanted to leave this here, if you still want to use Adv Markers with the MarkerClusterer then you can do the following
```
const mapMarker = new google.maps.marker.AdvancedMarkerView({
  map,
  position,
  content: marker,
});

mapMarker.getPosition = () => {
  return new google.maps.LatLng({
    lat: mapMarker.position.lat,
    lng: mapMarker.position.lng,
  });
};

mapMarker.setMap = (map) => (mapMarker.map = map);
mapMarker.setVisible = (v) => (mapMarker.visible = v);
mapMarker.getVisible = () =>
  mapMarker.visible === undefined ? true : mapMarker.visible || false;
```

Now this is just a hack until we get official support for this, but can help to get the job temporarily done

**Comment by ojg15** [comment]
Any update on this?

**Comment by nemanjacosovic** [comment]
@amuramoto @wangela could you please provide some input?

**Comment by TrevorMW** [comment]
PR is currently open that does compile and work. It's hacky, I know, but I needed to get this working for a client. Im working off the principle that AdvancedMarker class, while nowhere near the normal Marker class, does share most traits in one way or another, and the needed positional and map data can be pulled from AND set on the AdvancedMarker object.

**Comment by amitojha01** [comment]
Any update on this?

**Comment by TrevorMW** [comment]
@nemanjacosovic were you able to solve the issue of multiple clusters being added to the map on top of each other, slowing down the map? 

If so, would you mind I see your fixes? im still trying to figure that part out

**Comment by nemanjacosovic** [comment]
@TrevorMW I'm maybe misunderstanding your question, so I'm going to approach it as they were two different questions:

### Multiple clusters
I had no slow visible slowdown with multiple clusters, as they would never overlap each other.
The setting I've used was the `SuperClusterAlgorithm` and the settings of:
```
{ minPoints: 0, radius: 120 }
```

### Multiple markers
The approach I've took was to handle `MarkerClusterer` to render as-is, meaning. the clusters and the regular markers.
The initial parsing of data model (becoming an array of objects) for the markers allowed me to store a unique identifier, which I've used in parsing out both regular and advanced markers, and removing/adding them based on the zoom level.
The parsed model would be stored in a variable which was used for the `Marker` builder.

There is a 2 step process.

The **regular markers** would always be hidden by initially setting the `icon` property for the `Marker` as:
```
icon: {
  scaledSize: new google.maps.Size(0, 0),
  url: '',
},
```

After the **regular markers** would be rendered, a callback would then parse the DOM rendered markers to check which one of them has been used for the clusterer and which were placed as a pin on the map. The ones that were placed as the pin, would be stored (in the new/different variable).

This stored "visible markers" output would be used for building the `AdvancedMarkerView` markers by iterating over the model. There is also a clean up function that handles the zooming in and out, removing only the "re-clustered" markers.

So, the important thing was the sequence and optimisation by avoiding overhead as much as possible.
The slow down happened, in my case, where the rendered `AdvancedMarkerView` has a CSS animation which is affecting the map panning, but zooming in and out, works moderately fast. If you have "static" `AdvancedMarkerView` elements on the map, there shouldn't be any noticeable speed impact.

Apologies for not posting the code, as I'm unable to do so.

**Comment by VikashFluxon** [comment]
Any update on this issue? When is the ETA for release?

**Comment by amuramoto** [comment]
Very happy to let ya'll know we have a beta ready to test! You can install `@googlemaps/markerclusterer@beta`.

Please try it, break it, and let us know if you have feedback. We'll release this as `@latest` once we feel comfortable that it's stable.

**Comment by jisuo** [comment]
I tried the beta and I get `marker.getPosition is not a function` error as soon as I load the cluster. @amuramoto

**Comment by coder6652** [comment]
Is there a CDN for @googlemaps/markerclusterer@beta?

**Comment by amuramoto** [comment]
> Is there a CDN for @googlemaps/markerclusterer@beta?

https://unpkg.com/@googlemaps/markerclusterer@2.1.0-beta.1/dist/index.min.js

**Comment by amuramoto** [comment]
Just released `@2.1.0-beta.2` which migrates `AdvancedMarkerView` to `AdvancedMarkerElement` and `PinView` to `PinElement` which is the new class naming that is now available in the `weekly` channel.

**Comment by vicb** [comment]
I think this issue can be closed? (Advanced Markers are supported in the last release)
